File: v2/src/Inf_Tech_lab_Client/datarow.cpp
#include "datarow.h"
#include<bits/stdc++.h>
#include<QJsonDocument>
#include<QJsonObject>
#include<QDebug>
#include<datatypes.h>
#include<additional_functions.h>

using namespace std;

DataRow::DataRow(vector<DataTypes> & types , vector<pair<int , QString>> & _data)
{
    for(int i = 0 ; i < _data.size() ; ++i){
        data.push_back({_data[i].first , Value(types[i] , _data[i].second)});
    }
    is_sorted = false;
}

DataRow::DataRow(QJsonObject obj){
    for(auto e : obj.keys()){
        int key = e.toInt();
        data.push_back({key , Value(obj.value(e).toObject())});
    }
    is_sorted = false;
}

DataRow::DataRow(DataRow other , set<int> projection){
    for(auto & e : other.data){
        if(projection.find(e.first) != projection.end()){
            this->data.push_back(e);
        }
    }
    is_sorted = false;
}

QJsonValue DataRow::toJson()
{
    QJsonObject obj;
    for(auto e : data){
        obj.insert(QString::number(e.first) , QJsonValue(e.second.toJson()));
    }
    QJsonValue doc(obj);
    return doc;
}

void DataRow::save(QString & filename)
{
 //   saveJson(this->toJson() , filename);
}


vector<pair<int , DataRow::Value>> DataRow::getData()const{
    return this->data;
}

void DataRow::update(int pos , QString val){
    for(int i = 0 ; i < data.size() ; ++i){
        if(data[i].first == pos){
            data[i].second.value = val;
        }
    }
}

void DataRow::sortData(){
    if(is_sorted)return;
    sort(data.begin() , data.end() ,
        [](const pair<int , Value> & a, const pair<int , Value> & b) -> bool
        {
            return a.first < b.first;
        }
    );
    is_sorted = true;
}

bool DataRow::operator==(DataRow & other){
    if(this->data.size() != other.data.size()){
        return false;
    }

    sortData();
    other.sortData();

    bool res = 1;
    for(int i = 0 ; i < data.size() ; ++i){
        res &= (data[i].first == other.data[i].first);
        res &= (data[i].second.type == other.data[i].second.type);
        res &= (data[i].second.value == other.data[i].second.value);
    }
    return res;
}

DataRow DataRow::concat(DataRow row1 , DataRow row2 , map<int , int> code){
    for(auto &e : row2.data){
        row1.data.push_back({code[e.first] , e.second});
    }
    return row1;
}

File: 7/Inf_Tech_Server/DataRow.py
from DataTypes import *
import json

class Value:
    def __init__(self, _t, _v):
        self.type = _t
        self.value = _v

    def __init__(self, obj):
#       obj = dict(json.loads(obj))
        self.type = DataTypes.to_data_types(obj["type"])
        self.value = str(obj["value"])

    def toJson(self):
        dic = dict()
        dic["type"] = DataTypes.to_string(self.type)
        dic["value"] = self.value
        return dic

class DataRow:

    def __init__(self, types, _data):
        self.data = []
        for i in range(0, len(_data)):
            self.data.append([_data[i][0], Value(types[i], _data[i][1])])
        self.is_sorted = False

    def __init__(self, other, projection):
        self.data = []
        for e in other.data:
            if e[0] in projection:
                self.data.append(e)
        self.is_sorted = False

    def __init__(self, obj):
        self.is_sorted = False
        self.data = list()
        print(obj)
        obj = dict(obj)
        for e in obj.keys():
            key = int(e)
            self.data.append([key, Value(obj[e])])
        self.is_sorted = False

    def toJson(self):
        dic = dict()
        for e in self.data:
            dic[str(e[0])] = e[1].toJson()
        return dic

    def getData(self):
        return self.data

    def update(self, pos, val):
        for i in range(0, len(self.data)):
            if self.data[i][0] == pos:
                self.data[i][1] = val

    def sortData(self):
        if self.is_sorted:
            return
        self.data.sort(key = lambda e: e[0])
        self.is_sorted = True

    def __eq__(self, other):
        if len(self.data) != len(other.data):
            return False
        self.sortData()
        other.sortData()

        res = True
        for i in range(0, len(self.data)):
            res = res and self.data[i][0] == other.data[i][0]
            res = res and self.data[i][1].type == other.data[i][1].type
            res = res and self.data[i][1].value == other.data[i][1].value
        return res
